packages-exp/auth-exp/src/core/user/additional_user_info.test.ts

```diff
@@ -0,0 +1,152 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect } from 'chai';
+import { IdTokenResponse } from '../../model/id_token';
+import { ProviderId } from '../providers';
+import { fromIdTokenResponse } from './additional_user_info';
+
+describe('core/user/additional_user_info', () => {
+
+  describe('fromIdTokenResponse', () => {
+    const profile = {login: 'scott', friends: [], netWorth: 5.00};
+    const rawUserInfo = JSON.stringify(profile);
+
+    describe('parses federated IDP response tokens', () => {
+
+      it('for FacebookAdditionalUserInfo', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK, rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.FACEBOOK);
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+
+      it('for GithubAdditionalUserInfo', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.GITHUB, rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.GITHUB);
+        expect(username).to.eq('scott');
+        expect(profile).to.eq(profile);
+      });
+
+      it('for GoogleAdditionalUserInfo', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.GOOGLE, rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.GOOGLE);
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+
+      it('for TwitterAdditionalUserInfo', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.TWITTER, rawUserInfo, screenName: 'scott'}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.TWITTER);
+        expect(username).to.eq('scott');
+        expect(profile).to.eq(profile);
+      });
+    });
+
+    describe('parses profile data', () => {
+
+      it('for valid JSON', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK, rawUserInfo}))!.profile).to.deep.eq(profile);
```

Comment:
nit: use variables to avoid deep function nesting, this one in particular is a bit hard to read